### ISSUE TITLE
support multiple health checks per cluster

### ISSUE THREAD

**Issue by bdecoste** [kumahq/kuma#1927]
### Summary

If I configure a TCP and HTTP health checks for the same kuma service(envoy cluster) only the TCP health check is configrured

### Steps To Reproduce

```
$ kubectl -n kong-mesh-system get healthchecks -o yaml
apiVersion: v1
items:
- apiVersion: kuma.io/v1alpha1
  kind: HealthCheck
  mesh: default
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"kuma.io/v1alpha1","kind":"HealthCheck","mesh":"default","metadata":{"annotations":{},"name":"echo-service"},"spec":{"conf":{"alwaysLogHealthCheckFailures":true,"eventLogPath":"/dev/stdout","healthyPanicThreshold":"0","healthyThreshold":1,"http":{"path":"/get"},"interval":"10s","timeout":"20s","unhealthyThreshold":1},"destinations":[{"match":{"kuma.io/service":"echo-service_default_svc_8080"}}],"sources":[{"match":{"kuma.io/service":"*"}}]}}
    creationTimestamp: "2021-04-30T15:56:56Z"
    generation: 3
    managedFields:
    - apiVersion: kuma.io/v1alpha1
      fieldsType: FieldsV1
      fieldsV1:
        f:mesh: {}
        f:metadata:
          f:annotations:
            .: {}
            f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:spec:
          .: {}
          f:conf:
            .: {}
            f:alwaysLogHealthCheckFailures: {}
            f:eventLogPath: {}
            f:healthyPanicThreshold: {}
            f:healthyThreshold: {}
            f:http:
              .: {}
              f:path: {}
            f:interval: {}
            f:timeout: {}
            f:unhealthyThreshold: {}
          f:destinations: {}
          f:sources: {}
      manager: kubectl-client-side-apply
      operation: Update
      time: "2021-04-30T16:04:46Z"
    name: echo-service
    ownerReferences:
    - apiVersion: kuma.io/v1alpha1
      kind: Mesh
      name: default
      uid: d1241d07-e67c-4f1f-83f9-50056379473d
    resourceVersion: "22844071"
    selfLink: /apis/kuma.io/v1alpha1/healthchecks/echo-service
    uid: fd8bcf8d-b048-4a81-abf2-87517545fda1
  spec:
    conf:
      alwaysLogHealthCheckFailures: true
      eventLogPath: /dev/stdout
      healthyPanicThreshold: "0"
      healthyThreshold: 1
      http:
        path: /get
      interval: 10s
      timeout: 20s
      unhealthyThreshold: 1
    destinations:
    - match:
        kuma.io/service: echo-service_default_svc_8080
    sources:
    - match:
        kuma.io/service: '*'
- apiVersion: kuma.io/v1alpha1
  kind: HealthCheck
  mesh: default
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"kuma.io/v1alpha1","kind":"HealthCheck","mesh":"default","metadata":{"annotations":{},"name":"tcp-echo"},"spec":{"conf":{"alwaysLogHealthCheckFailures":true,"eventLogPath":"/dev/stdout","healthyPanicThreshold":"0","healthyThreshold":1,"interval":"2s","tcp":{},"timeout":"20s","unhealthyThreshold":1},"destinations":[{"match":{"kuma.io/service":"echo-service_default_svc_8080"}}],"sources":[{"match":{"kuma.io/service":"*"}}]}}
    creationTimestamp: "2021-04-30T16:08:47Z"
    generation: 1
    managedFields:
    - apiVersion: kuma.io/v1alpha1
      fieldsType: FieldsV1
      fieldsV1:
        f:mesh: {}
        f:metadata:
          f:annotations:
            .: {}
            f:kubectl.kubernetes.io/last-applied-configuration: {}
        f:spec:
          .: {}
          f:conf:
            .: {}
            f:alwaysLogHealthCheckFailures: {}
            f:eventLogPath: {}
            f:healthyPanicThreshold: {}
            f:healthyThreshold: {}
            f:interval: {}
            f:tcp: {}
            f:timeout: {}
            f:unhealthyThreshold: {}
          f:destinations: {}
          f:sources: {}
      manager: kubectl-client-side-apply
      operation: Update
      time: "2021-04-30T16:08:47Z"
    name: tcp-echo
    ownerReferences:
    - apiVersion: kuma.io/v1alpha1
      kind: Mesh
      name: default
      uid: d1241d07-e67c-4f1f-83f9-50056379473d
    resourceVersion: "22844924"
    selfLink: /apis/kuma.io/v1alpha1/healthchecks/tcp-echo
    uid: 336a9a70-0c58-4cf7-9f4d-7f9453954e16
  spec:
    conf:
      alwaysLogHealthCheckFailures: true
      eventLogPath: /dev/stdout
      healthyPanicThreshold: "0"
      healthyThreshold: 1
      interval: 2s
      tcp: {}
      timeout: 20s
      unhealthyThreshold: 1
    destinations:
    - match:
        kuma.io/service: echo-service_default_svc_8080
    sources:
    - match:
        kuma.io/service: '*'
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

```
    {
     "version_info": "6c918c8f-45ab-4ffc-9d26-d7be75b15a9c",
     "cluster": {
      "@type": "type.googleapis.com/envoy.config.cluster.v3.Cluster",
      "name": "echo-service_default_svc_8080",
      "type": "EDS",
      "eds_cluster_config": {
       "eds_config": {
        "ads": {},
        "resource_api_version": "V3"
       }
      },
      "connect_timeout": "5s",
      "health_checks": [
       {
        "timeout": "20s",
        "interval": "2s",
        "unhealthy_threshold": 1,
        "healthy_threshold": 1,
        "tcp_health_check": {},
        "event_log_path": "/dev/stdout",
        "always_log_health_check_failures": true
       }
      ],
```

**Comment by subnetmarco** [comment]
@lobkovilya thoughts?

**Comment by jakubdyszkiewicz** [comment]
Hey,

please take a look [how Kuma maches policies](https://kuma.io/docs/1.1.4/policies/how-kuma-chooses-the-right-policy-to-apply/). Only the most specific one matches.

It's true that right now we have a validation that `tcp` and `http` section is `oneof`, so you can pick only one of them in one policy.
I see that cluster supports many healtcheck so we could get rid of this validation and if user specifies `tcp` and `http` it will pass.

@bartsmykla is there any specific reason that we used `oneof`?

**Comment by bartsmykla** [comment]
I don't think there is a valid reason, probably only my assumption the only one is valid at a time during implementation

I currently don't see any reason we couldn't get rid of this "requirement"

**Comment by austince** [comment]
I think this was closed by #1951, please re-open if not.
